pkg/catalog/catalog.go

```diff
@@ -1763,6 +1816,192 @@ func (c *Catalog) GetRange(ctx context.Context, repositoryID, rangeID string) (g
 	return c.Store.GetRange(ctx, repository, graveler.RangeID(rangeID))
 }
 
+func (c *Catalog) ensureBranchExists(ctx context.Context, repository *graveler.RepositoryRecord, branch, sourceBranch string) error {
+	branchID := graveler.BranchID(branch)
+	_, err := c.Store.GetBranch(ctx, repository, branchID)
+	if errors.Is(err, graveler.ErrNotFound) {
+		_, err = c.Store.CreateBranch(ctx, repository, branchID, graveler.Ref(sourceBranch))
+	}
+	return err
+}
+
+func (c *Catalog) importAsync(repository *graveler.RepositoryRecord, branchID, importID string, params ImportRequest) {
+	ctx, cancel := context.WithCancel(context.Background()) // Need a new context for the async operations
+	defer cancel()
+	logger := c.log.WithField("import_id", importID)
+	importManager, err := NewImport(ctx, cancel, logger, c.KVStore, repository, importID)
+	if err != nil {
+		logger.WithError(err).Error("creating import manager")
+		return
+	}
+	defer importManager.Close()
+
+	importStatus := importManager.Status()
+	importBranch := "_" + branchID + "_imported"
+	err = c.ensureBranchExists(ctx, repository, importBranch, branchID)
+	if err != nil {
+		importStatus.Error = fmt.Errorf("ensure import branch: %w", err)
+		importManager.StatusChan <- importStatus
+		return
+	}
+	importStatus.ImportBranch = &importBranch
+
+	var wg multierror.Group
+	const ingestChanSize = 10
+	ingestChan := make(chan *walkEntryIterator, ingestChanSize)
+	wg.Go(func() error {
+		for i := range ingestChan {
+			it := i // Pinning
+			wg.Go(func() error {
+				defer it.Close()
+				return importManager.Ingest(it)
+			})
+		}
+		return nil
+	})
+
+	for _, source := range params.Paths {
+		// TODO (niro): Need to handle this at some point (use adapter GetWalker)
+		walker, err := c.walkerFactory.GetWalker(ctx, store.WalkerOptions{StorageURI: source.Path})
+		if err != nil {
+			importStatus.Error = fmt.Errorf("creating object-store walker on path %s: %w", source.Path, err)
+			importManager.StatusChan <- importStatus
+			return
+		}
+
+		it, err := NewWalkEntryIterator(ctx, walker, source.Type, source.Destination, "", "")
+		if err != nil {
+			importStatus.Error = fmt.Errorf("creating walk iterator on path %s: %w", source.Path, err)
+			importManager.StatusChan <- importStatus
+			return
+		}
+		logger.WithFields(logging.Fields{"source": source.Path, "itr": it}).Debug("Ingest source")
+		ingestChan <- it
```

Comment:
passing iterator we need to close over a channel is risky as we need to make sure we close each one.